linker/Linker.Steps/CodeRewriterStep.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Mono.Linker.Steps {
+	public class CodeRewriterStep : BaseStep
+	{
+		AssemblyDefinition assembly;
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			this.assembly = assembly;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			foreach (var method in type.Methods) {
+				if (method.HasBody)
+					ProcessMethod (method);
+			}
+
+			foreach (var nested in type.NestedTypes)
+				ProcessType (nested);
+		}
+
+		void ProcessMethod (MethodDefinition method)
+		{
+			switch (Annotations.GetAction (method)) {
+			case MethodAction.ConvertToStub:
+				RewriteBodyToStub (method);
+				break;
+			case MethodAction.ConvertToThrow:
+				RewriteBodyToLinkedAway (method);
+				break;
+			}
+		}
+
+		void RewriteBodyToLinkedAway (MethodDefinition method)
+		{
+			method.ImplAttributes &= ~(MethodImplAttributes.AggressiveInlining | MethodImplAttributes.Synchronized);
+			method.ImplAttributes |= MethodImplAttributes.NoInlining;
+
+			method.Body = CreateThrowLinkedAwayBody (method);
+			ClearDebugInformation (method);
+		}
+
+		void RewriteBodyToStub (MethodDefinition method)
+		{
+			method.Body = CreateStubBody (method);
+
+			ClearDebugInformation (method);
+		}
+
+		MethodBody CreateThrowLinkedAwayBody (MethodDefinition method)
+		{
+			var body = new MethodBody (method);
+			var il = body.GetILProcessor ();
+
+			// import the method into the current assembly
+			var ctor = Context.MarkedKnownMembers.NotSupportedExceptionCtorString;
+			ctor = assembly.MainModule.ImportReference (ctor);
+
+			il.Emit (OpCodes.Ldstr, "Linked away");
+			il.Emit (OpCodes.Newobj, ctor);
+			il.Emit (OpCodes.Throw);
+			return body;
+		}
+
+		MethodBody CreateStubBody (MethodDefinition method)
+		{
+			var body = new MethodBody (method);
+
+			if (method.HasParameters && method.Parameters.Any (l => l.IsOut))
+				throw new NotImplementedException ();
+
+			var il = body.GetILProcessor ();
+			if (method.IsConstructor && !method.IsStatic) {
+				var base_ctor = GetDefaultInstanceConstuctor (method.DeclaringType.BaseType);
+				base_ctor = assembly.MainModule.ImportReference (base_ctor);
+
+				il.Emit (OpCodes.Ldarg_0);
+				il.Emit (OpCodes.Call, base_ctor);
+			}
+
+			switch (method.ReturnType.MetadataType) {
+			case MetadataType.Void:
+				break;
+			case MetadataType.Boolean:
+				il.Emit (OpCodes.Ldc_I4_0);
+				break;
+			default:
+				throw new NotImplementedException (method.ReturnType.FullName);
+			}
+
+			il.Emit (OpCodes.Ret);
+			return body;
+		}
+
+		static MethodReference GetDefaultInstanceConstuctor (TypeReference type)
+		{
+			foreach (var m in type.GetMethods ()) {
+				if (m.HasParameters)
+					continue;
+
+				var definition = m.Resolve ();
```

Comment:
Can you add a null check to guard against a failed resolve just in case?

Reply:
This is after mark and sweep so the method must resolve, are you suggesting to throw on null?

Reply:
Good point.   I guess it's OK.  I was thinking of continuing if it was null.

I'll just have to be careful to put this step after our step that injects unresolved types in UnityLinker.  We override the `HandleUnresolvedType` methods in the MarkStep and track unresolved things instead of throwing.  Then a later step stubs back in missing things the best we can.